File: day09/main.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	rope1 := NewRope(1)
	rope2 := NewRope(9)

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		dir, countstr, ok := strings.Cut(scanner.Text(), " ")
		if !ok {
			log.Fatalln("Invalid line")
		}
		count, err := strconv.Atoi(countstr)
		if err != nil {
			log.Fatalln(err)
		}
		for i := 0; i < count; i++ {
			if err := rope1.Move(dir); err != nil {
				log.Fatalln(err)
			}
			if err := rope2.Move(dir); err != nil {
				log.Fatalln(err)
			}
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Part 1:", len(rope1.history))
	fmt.Println("Part 2:", len(rope2.history))
}

type (
	Tuple2 struct {
		y, x int
	}

	Rope struct {
		h       Tuple2
		t       []Tuple2
		history map[Tuple2]struct{}
	}
)

func (t Tuple2) Move(dir string) (Tuple2, error) {
	switch dir {
	case "U":
		t.y -= 1
	case "R":
		t.x += 1
	case "D":
		t.y += 1
	case "L":
		t.x -= 1
	default:
		return Tuple2{}, errors.New("Invalid direction")
	}
	return t, nil
}

func (t Tuple2) Delta(p Tuple2) Tuple2 {
	t.x += p.x
	t.y += p.y
	return t
}

func (t Tuple2) Dist(p Tuple2) Tuple2 {
	return Tuple2{
		x: p.x - t.x,
		y: p.y - t.y,
	}
}

func (t Tuple2) Dir() Tuple2 {
	if t.x != 0 {
		if t.x > 0 {
			t.x = 1
		} else {
			t.x = -1
		}
	}
	if t.y != 0 {
		if t.y > 0 {
			t.y = 1
		} else {
			t.y = -1
		}
	}
	return t
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func (t Tuple2) MaxMag() int {
	a := abs(t.x)
	b := abs(t.y)
	if a > b {
		return a
	}
	return b
}

func NewRope(size int) *Rope {
	history := map[Tuple2]struct{}{
		{}: {},
	}
	tail := make([]Tuple2, size)
	return &Rope{
		h:       Tuple2{},
		t:       tail,
		history: history,
	}
}

func (r *Rope) Move(dir string) error {
	next, err := r.h.Move(dir)
	if err != nil {
		return err
	}
	r.h = next
	last := len(r.t) - 1
	for i, t := range r.t {
		if k := t.Dist(next); k.MaxMag() > 1 {
			next = t.Delta(k.Dir())
			if i == last {
				r.history[next] = struct{}{}
			}
		} else {
			next = t
		}
		r.t[i] = next
	}
	return nil
}

File: day09/src/main.rs
use std::collections::HashSet;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

type BoxError = Box<dyn std::error::Error>;
type BoxResult<T> = Result<T, BoxError>;

fn main() -> BoxResult<()> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut rope1 = Rope::new(1);
    let mut rope2 = Rope::new(9);

    for line in reader.lines() {
        let line = line?;
        let (dir, countstr) = line
            .split_once(' ')
            .ok_or::<BoxError>("Invalid line".into())?;
        let count = countstr.parse::<i32>()?;
        for _ in 0..count {
            rope1.apply_dir(&dir)?;
            rope2.apply_dir(&dir)?;
        }
    }

    println!(
        "Part 1: {}\nPart 2: {}",
        rope1.history.len(),
        rope2.history.len()
    );

    Ok(())
}

#[derive(Clone, Copy, Eq, Hash, PartialEq)]
struct Tuple2 {
    x: i32,
    y: i32,
}

impl Tuple2 {
    fn apply_dir(&mut self, dir: &str) -> BoxResult<()> {
        match dir {
            "U" => self.y -= 1,
            "R" => self.x += 1,
            "D" => self.y += 1,
            "L" => self.x -= 1,
            _ => return Err("Invalid direction".into()),
        }
        Ok(())
    }

    fn delta(&mut self, p: &Self) {
        self.x += p.x;
        self.y += p.y;
    }

    fn dist(&self, p: &Self) -> Self {
        Self {
            x: p.x - self.x,
            y: p.y - self.y,
        }
    }

    fn dir(&self) -> Self {
        Self {
            x: unit_dir(self.x),
            y: unit_dir(self.y),
        }
    }

    fn max_mag(&self) -> i32 {
        self.x.abs().max(self.y.abs())
    }
}

fn unit_dir(a: i32) -> i32 {
    if a == 0 {
        0
    } else if a > 0 {
        1
    } else {
        -1
    }
}

struct Rope {
    h: Tuple2,
    t: Vec<Tuple2>,
    history: HashSet<Tuple2>,
}

impl Rope {
    fn new(size: usize) -> Self {
        let mut history = HashSet::new();
        history.insert(Tuple2 { x: 0, y: 0 });
        Self {
            h: Tuple2 { x: 0, y: 0 },
            t: vec![Tuple2 { x: 0, y: 0 }; size],
            history,
        }
    }

    fn apply_dir(&mut self, dir: &str) -> BoxResult<()> {
        self.h.apply_dir(dir)?;
        let mut next = &self.h;
        let last = self.t.len() - 1;
        for i in 0..self.t.len() {
            let k = self.t[i].dist(&next);
            if k.max_mag() > 1 {
                self.t[i].delta(&k.dir());
                if i == last {
                    self.history.insert(self.t[i]);
                }
            }
            next = &self.t[i];
        }
        Ok(())
    }
}
